Cubical/Displayed/Auto.agda

```diff
@@ -0,0 +1,361 @@
+{-
+
+  - Automatically generate UARel and DUARel instances
+
+-}
+{-# OPTIONS --cubical --no-import-sorts --no-exact-split --safe #-}
+module Cubical.Displayed.Auto where
+
+open import Cubical.Foundations.Prelude
+open import Cubical.Foundations.Equiv
+open import Cubical.Foundations.Function
+open import Cubical.Foundations.Isomorphism
+open import Cubical.Foundations.Univalence
+
+open import Cubical.Displayed.Base
+open import Cubical.Displayed.Subst
+open import Cubical.Displayed.Morphism
+
+open import Cubical.Displayed.Constant
+open import Cubical.Displayed.Function
+open import Cubical.Displayed.Generic
+open import Cubical.Displayed.Sigma
+open import Cubical.Displayed.Unit
+open import Cubical.Displayed.Universe
+
+open import Cubical.Data.List.Base
+open import Cubical.Data.Nat.Base
+open import Cubical.Data.Sigma.Base
+open import Cubical.Data.Unit.Base
+
+import Agda.Builtin.Reflection as R
+open import Cubical.Reflection.Base
+
+-- Descriptor language
+
+mutual
+  data UARelDesc : ∀ {ℓA ℓ≅A} {A : Type ℓA} (𝒮-A : UARel A ℓ≅A) → Typeω where
+
+    generic : ∀ {ℓA} {A : Type ℓA} → UARelDesc (𝒮-generic A)
+
+    univ : ∀ ℓU → UARelDesc (𝒮-Univ ℓU)
+
+    sigma : ∀ {ℓA ℓ≅A ℓB ℓ≅B}
+      {A : Type ℓA} {𝒮-A : UARel A ℓ≅A} (dA : UARelDesc 𝒮-A)
+      {B : A → Type ℓB} {𝒮ᴰ-B : DUARel 𝒮-A B ℓ≅B} (dB : DUARelDesc 𝒮-A 𝒮ᴰ-B)
+      → UARelDesc (∫ 𝒮ᴰ-B)
+
+    param : ∀ {ℓA ℓB ℓ≅B} (A : Type ℓA)
+      {B : Type ℓB} {𝒮-B : UARel B ℓ≅B} (dB : UARelDesc 𝒮-B)
+      → UARelDesc (A →𝒮 𝒮-B)
+
+    pi : ∀ {ℓA ℓ≅A ℓB ℓ≅B}
+      {A : Type ℓA} {𝒮-A : UARel A ℓ≅A} (dA : UARelDesc 𝒮-A)
+      {B : A → Type ℓB} {𝒮ᴰ-B : DUARel 𝒮-A B ℓ≅B} (dB : DUARelDesc 𝒮-A 𝒮ᴰ-B)
+      → UARelDesc (𝒮-Π 𝒮-A 𝒮ᴰ-B)
+
+    unit : UARelDesc 𝒮-Unit
+
+  -- Projections from one UARel to another
+  data UARelReindex : ∀ {ℓA ℓ≅A ℓC ℓ≅C}
+    {A : Type ℓA} {𝒮-A : UARel A ℓ≅A}
+    {C : Type ℓC} {𝒮-C : UARel C ℓ≅C}
+    (f : UARelHom 𝒮-A 𝒮-C)
+    → Typeω
```

Comment:
Why is this `Typeω`?

Reply:
Because the levels are indexes I would guess, do we need that?

Reply:
Oh right!

Reply:
They're indices because the level of a compound type expression is built up from the levels of its components, eg. the level of a sigma is the max of the domain and codomain level. I couldn't see away around this.